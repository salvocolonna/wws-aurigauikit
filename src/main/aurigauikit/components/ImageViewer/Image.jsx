import React from "react"
import sizeMe from "react-sizeme"

const refreshRate = 16

function throttledEvent(name, refreshRate = 16) {
  let running = false
  const run = () => {
    window.dispatchEvent(new CustomEvent(name))
    running = false
  }
  return () => {
    if (running) {
      return
    }
    running = true
    if (refreshRate === 16) requestAnimationFrame(run)
    else setTimeout(run, refreshRate)
  }
}

const throttleResize = throttledEvent("optimizedResize")

export default sizeMe({ monitorHeight: true, refreshRate })(
  class extends React.Component {
    state = { loaded: false }

    resize = () => {
      if (this.img && this.props.currentSize) {
        const width = this.img.offsetWidth
        const height = this.img.offsetHeight
        if (width !== this.props.currentSize.width || height !== this.props.currentSize.height)
          this.props.onResize({ width, height })
      }
    }

    shouldComponentUpdate({ src, currentSize }, { loaded }) {
      return (
        this.props.src !== src ||
        this.props.currentSize !== currentSize ||
        loaded !== this.state.loaded
      )
    }

    componentDidMount() {
      if (this.img) this.img.onload = () => this.setState({ loaded: true }, this.resize)
      window.addEventListener("resize", throttleResize)
      window.addEventListener("optimizedResize", this.resize)
      this.resize()
    }

    componentWillUnmount() {
      window.removeEventListener("optimizedResize", this.resize)
      window.removeEventListener("resize", throttleResize)
    }

    render() {
      const { src, style, className, bordered, circle } = this.props
      const borders = {
        border: bordered ? "1px solid #2984C5" : undefined,
        borderRadius: circle ? "50%" : bordered ? 6 : undefined,
        objectFit: circle ? "cover" : undefined
      }
      return (
        <img
          ref={img => (this.img = img)}
          src={src}
          style={{
            ...borders,
            ...style,
            visibility: !this.state.loaded ? "hidden" : undefined
          }}
          className={className}
        />
      )
    }
  }
)
