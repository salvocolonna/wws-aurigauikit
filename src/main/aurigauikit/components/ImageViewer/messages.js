import { defineMessages } from "react-intl"

const PATH = "image-viewer"

export default defineMessages({
  upload: {
    id: `${PATH}.upload`,
    defaultMessage: "Upload..."
  },
  edit: {
    id: `${PATH}.edit`,
    defaultMessage: "Edit"
  }
})
