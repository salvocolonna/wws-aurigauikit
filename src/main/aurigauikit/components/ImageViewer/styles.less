
.ImageViewer{
  pointer-events: none;
  text-align: center;
  img{ 
    pointer-events: all; 
    object-fit: contain;
  }
  .ImageViewer-overlay{
    pointer-events: all;
    margin: auto;
    opacity: 0;
    transition: opacity .5s ease;
    background-color: rgba(41, 132, 197,.2);
    transform: translateY(-100%);
    box-shadow: 0px 0px 10px #2984C5;
		display: flex;
		align-items: center;
		justify-content: center;
    cursor: pointer;
    overflow: hidden;
    .ImageViewer-uploader{
      font-weight: bold; 
      font-size: 20px;
      color: #fff;
      background-color: #2984C5;
      display: flex;
      align-items: center;
      justify-content: center;
      width: 100%; 
      height: 30%;
    }
  }
  &:hover{
    .ImageViewer-overlay {
      opacity: 1;
    }
  }
}

