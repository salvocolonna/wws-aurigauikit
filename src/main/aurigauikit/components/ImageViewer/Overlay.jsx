import React from "react"
import { FormattedMessage as Msg } from "react-intl"
import styles from "./styles.less"
import messages from "./messages"
import sizeMe from "react-sizeme"
import ContextMenu from "aurigauikit/components/ContextMenu"

export default sizeMe({
  monitorHeight: true,
  monitorPosition: true
})(
  class extends React.Component {
    constructor(props) {
      super(props)
      this.container = React.createRef()
      this.state = { menu: { context: { position: null } } }
    }

    closeMenu = () => this.setState({ menu: { context: { position: null } } })

    openMenu = e => {
      const {
        width,
        height,
        position: { top, left }
      } = this.props.size
      const position = {
        left: this.container.current.offsetLeft + width / 2,
        top: this.container.current.offsetTop - height / 2
      }
      this.setState({ menu: { context: { position } } })
    }

    click = () => {
      const { onClick, editable } = this.props
      setTimeout(this.closeMenu, 0)
      this.closeMenu()
      if (editable && onClick) onClick()
    }

    render() {
      const { editable, width, height, bordered, circle, className, style, onRemove } = this.props
      return (
        <div
          className={["ImageViewer-overlay", className].join(" ")}
          style={{
            width,
            height,
            marginBottom: -height,
            borderRadius: circle ? "50%" : bordered ? 6 : undefined,
            ...style
          }}
          onClick={onRemove ? this.openMenu : this.click}
          ref={this.container}>
          {onRemove && (
            <div className={"ImageViewer-uploader"}>
              <i style={{ marginRight: 10 }} className="fa fa-pencil" />
              <Msg {...messages.edit} />
              <ContextMenu
                center
                items={[
                  {
                    title: "Upload",
                    action: this.click,
                    iconName: "upload"
                  },
                  {
                    title: "Remove",
                    action: () => editable && onRemove && onRemove(),
                    iconName: "remove",
                    style: "destructive"
                  }
                ]}
                onCloseRequested={this.closeMenu}
                context={this.state.menu.context}
              />
            </div>
          )}
        </div>
      )
    }
  }
)
