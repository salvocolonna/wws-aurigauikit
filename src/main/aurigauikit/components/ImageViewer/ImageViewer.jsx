import React from "react"
import "./styles.less"
import emptyImage from "./empty-image.png"

import Image from "./Image"
import Overlay from "./Overlay"

export default class extends React.Component {
  state = { size: { width: 0, height: 0 } }

  sizeUpdated = size => this.setState({ size })

  imageUploaded = file => {
    const { onUpload } = this.props
    if (onUpload) {
      const src = URL.createObjectURL(file)
      onUpload(src)
    }
  }

  render() {
    const { size } = this.state
    const {
      editable,
      imageStyle,
      className,
      imageClassName,
      style,
      bordered,
      circle,
      canRemove,
      onRemove
    } = this.props
    const classNames = ["ImageViewer", className]
    return (
      <div style={style} className={classNames.join(" ")}>
        <Image
          onSize={this.sizeUpdated}
          onResize={this.sizeUpdated}
          style={imageStyle}
          currentSize={this.state.size}
          className={imageClassName}
          src={this.props.src || emptyImage}
          bordered={bordered}
          circle={circle}
        />
        {editable && (
          <Overlay
            canRemove={canRemove}
            onRemove={onRemove}
            editable={editable}
            width={size.width}
            height={size.height}
            bordered={bordered}
            style={imageStyle}
            className={imageClassName}
            onClick={() => this.uploader && this.uploader.click()}
            circle={circle}
          />
        )}
        {editable && (
          <input
            ref={uploader => (this.uploader = uploader)}
            onChange={e => this.imageUploaded(e.target.files[0])}
            type="file"
            style={{ display: "none" }}
          />
        )}
      </div>
    )
  }
}
