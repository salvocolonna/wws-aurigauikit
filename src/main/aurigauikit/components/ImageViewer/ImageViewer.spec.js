import React from "react"
import renderer from "react-test-renderer"

import ImageViewer from "./ImageViewer"

jest.mock("./Image")
jest.mock("./Overlay")

describe("ImageViewer", () => {
  it("displays as expected", () => {
    const component = renderer.create(<ImageViewer />)
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it("forewards style and className", () => {
    const component = renderer.create(
      <ImageViewer imageStyle={{ borderRadius: 50 }} className="image-viewer" />
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it("displays as expected editable", () => {
    const component = renderer.create(<ImageViewer editable />)
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
