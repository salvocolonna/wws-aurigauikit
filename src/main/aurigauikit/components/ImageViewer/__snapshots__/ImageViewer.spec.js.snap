// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageViewer displays as expected 1`] = `
<div
  className="ImageViewer "
>
  Image {"currentSize":{"width":0,"height":0},"src":"mock/resource/src"}
</div>
`;

exports[`ImageViewer displays as expected editable 1`] = `
<div
  className="ImageViewer "
>
  Image {"currentSize":{"width":0,"height":0},"src":"mock/resource/src"}
  Overlay {"editable":true,"width":0,"height":0}
  <input
    onChange={[Function]}
    style={
      Object {
        "display": "none",
      }
    }
    type="file"
  />
</div>
`;

exports[`ImageViewer forewards style and className 1`] = `
<div
  className="ImageViewer image-viewer"
>
  Image {"style":{"borderRadius":50},"currentSize":{"width":0,"height":0},"src":"mock/resource/src"}
</div>
`;
